fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingScrubber.java

```diff
@@ -0,0 +1,359 @@
+                        .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.SCRUB_REPAIR)
+                        .build();
+    }
+
+    @Nonnull
+    private OnlineIndexer.ScrubbingPolicy spolicy() {
+        assert (scrubbingPolicy != null);// this can never happen, just eliminating compiler warnings
+        return scrubbingPolicy;
+    }
+
+    @Override
+    CompletableFuture<Void> buildIndexInternalAsync() {
+        return getRunner().runAsync(context -> openRecordStore(context)
+                .thenCompose( store ->
+                        context.getReadVersionAsync()
+                                .thenCompose(vignore -> {
+                                    SubspaceProvider subspaceProvider = common.getRecordStoreBuilder().getSubspaceProvider();
+                                    return subspaceProvider.getSubspaceAsync(context)
+                                            .thenCompose(subspace ->
+                                                    scrubIndex(subspaceProvider, subspace, null, null)
+                                                    .thenCompose(ignore -> scrubRecords(subspaceProvider, subspace, null, null))
+                                            );
+                                })
+        ), common.indexLogMessageKeyValues("IndexingScrubber::buildIndexInternalAsync"));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> scrubIndex(@Nonnull SubspaceProvider subspaceProvider, @Nonnull Subspace subspace,
+                                               @Nullable byte[] start, @Nullable byte[] end) {
+        if (!spolicy().shouldScrubDangling()) {
+            return AsyncUtil.DONE;
+        }
+
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "scrubRecords",
+                LogMessageKeys.RANGE_START, start,
+                LogMessageKeys.RANGE_END, end);
+
+        return iterateAllRanges(additionalLogMessageKeyValues,
+                (store, recordsScanned) -> scrubIndexRangeOnly(store, start, end, recordsScanned),
+                subspaceProvider, subspace);
+    }
+
+    @Nonnull
+    private CompletableFuture<Boolean> scrubIndexRangeOnly(@Nonnull FDBRecordStore store, byte[] startBytes, byte[] endBytes, @Nonnull AtomicLong recordsScanned) {
+        // return false when done
+        Index index = common.getIndex();
+        final RecordMetaData metaData = store.getRecordMetaData();
+        final RecordMetaDataProvider recordMetaDataProvider = common.getRecordStoreBuilder().getMetaDataProvider();
+        if (recordMetaDataProvider == null || !metaData.equals(recordMetaDataProvider.getRecordMetaData())) {
+            throw new MetaDataException("Store does not have the same metadata");
+        }
+        final IndexMaintainer maintainer = store.getIndexMaintainer(index);
+
+        // scrubbing idempotence indexes only (for now)
+        validateOrThrowEx(maintainer.isIdempotent(), "scrubbed index is not idempotent");
```

Comment:
This feels like not quite the right check. Maybe it'll come out in the wash when the index is actually scanned, but I think the thing to check is that the index type supports scanning `BY_VALUE`.

At the moment, there's not a terrific way of knowing if an index type supports scanning `BY_VALUE`. We could theoretically add that as a method on the index maintainer, I suppose.

Reply:
Should this new maintainer method be part of this PR?